noisy-simulator/src/state_vector_simulator.rs

```diff
@@ -0,0 +1,359 @@
+
+    /// Returns the number of qubits in the system.
+    pub fn number_of_qubits(&self) -> usize {
+        self.number_of_qubits
+    }
+
+    /// Returns `true` if the squared L2 norm of the matrix is 1.
+    fn is_normalized(&self) -> bool {
+        (self.norm_squared() - 1.0).abs() <= TOLERANCE
+    }
+
+    /// Returns the squared L2 norm of the matrix.
+    fn norm_squared(&self) -> f64 {
+        self.data.norm_squared()
+    }
+
+    /// Return theoretical change in trace due to operations that have been applied so far.
+    /// In reality, the density matrix is always renormalized after instruments / operations
+    /// have been applied.
+    pub fn trace_change(&self) -> f64 {
+        self.trace_change
+    }
+
+    /// Renormalizes the matrix such that the trace is 1.
+    fn renormalize(&mut self) -> Result<(), Error> {
+        self.renormalize_with_norm_squared(self.norm_squared())
+    }
+
+    /// Renormalizes the matrix such that the trace is 1. Uses a precomputed `norm_squared`.
+    fn renormalize_with_norm_squared(&mut self, norm_squared: f64) -> Result<(), Error> {
+        if norm_squared < TOLERANCE {
+            return Err(Error::ProbabilityZeroEvent);
+        }
+        let renormalization_factor = 1.0 / norm_squared.sqrt();
+        for entry in self.data.iter_mut() {
+            *entry *= renormalization_factor;
+        }
+        Ok(())
+    }
+
+    /// Return the probability of a given effect.
+    fn effect_probability(
+        &self,
+        effect_matrix: &SquareMatrix,
+        qubits: &[usize],
+    ) -> Result<f64, Error> {
+        let mut state_copy = self.data.clone();
+        apply_kernel(&mut state_copy, effect_matrix, qubits)?;
+        Ok(state_copy.dot(&self.data.conjugate()).re)
+    }
+
+    fn sample_kraus_operators(
+        &mut self,
+        kraus_operators: &[SquareMatrix],
+        qubits: &[usize],
+        renormalization_factor: f64,
+        random_sample: f64,
+    ) -> Result<(), Error> {
+        let mut summed_probability = 0.0;
+        let mut last_non_zero_probability = 0.0;
+        let mut last_non_zero_probability_index = 0;
+
+        for (i, kraus_operator) in kraus_operators.iter().enumerate() {
+            let mut state_copy = self.data.clone();
+            apply_kernel(&mut state_copy, kraus_operator, qubits)?;
+            let norm_squared = state_copy.norm_squared();
+            let p = norm_squared / renormalization_factor;
+            summed_probability += p;
+            if p >= TOLERANCE {
+                last_non_zero_probability = p;
+                last_non_zero_probability_index = i;
+                if summed_probability > random_sample {
+                    self.data = state_copy;
+                    self.renormalize_with_norm_squared(norm_squared)?;
+                    return Ok(());
+                }
+            }
+        }
+        if summed_probability + TOLERANCE > random_sample && last_non_zero_probability >= TOLERANCE
+        {
```

Comment:
I think cargo fmt needs to be run on this.

Reply:
I tried that but then the formatter that we have on our repo (what formats the code when you hit Ctrl + S) automatically formats it that way each time I save the file.